specs/derivation.md

```diff
@@ -0,0 +1,820 @@
+
+### L1 Traversal
+
+In the *L1 Traversal* stage, we simply read the header of the next L1 block. In normal operations, these will be new
+L1 blocks as they get created, though we can also read old blocks while syncing, or in case of an L1 [re-org][g-reorg].
+
+### L1 Retrieval
+
+In the *L1 Retrieval* stage, we read the block we get from the outer stage (L1 traversal), and extract data for it. In
+particular we extract a byte string that corresponds to the concatenation of the data in all the [batcher
+transaction][g-batcher-transaction] belonging to the block. This byte stream encodes a stream of [channel
+frames][g-channel-frame] (see the [Batch Submission Wire Format][wire-format] section for more info).
+
+This frames are parsed, then grouped them per [channel][g-channel] into a structure we call the *channel bank*.
+
+Some frames are ignored:
+
+- Frames where `frame.frame_number <= progress`, where `progress` is the number of the latest frame that was read.
+    - i.e. in case of duplicate frame, the first frame read from L1 is considered canonical.
+- Frames past the end of the channel (i.e. past the first frame marked with `frame.is_last == 1`) are ignored.
+    - These frames could still be written into the channel bank if we haven't seen the last frame yet. But they will
+      never be read out from the channel bank.
+
+### Channel Bank
+
+The *Channel Bank* stage is responsible from reading from the channel bank that was written to by the L1 retrieval
+stage, and decompressing batches from these frames.
+
+In principle, we should be able to read any channel that has any number of sequential frames at the "front" of the
+channel (i.e. right after any frames that have been read from the bank already) and decompress batchers from them. (Note
+that if we did this, we'd need to keep partially decompressed batches around.)
+
+However, our current implementation doesn't support streaming decompression, so currently we have to wait until either:
+
+- We have received all frames in the channel (i.e. we received the last frame in the channel (`is_last == 1`) and every
+  frame before it)
+- The channel has timed out (in which we case we read all contiguous sequential frames from the start of the channel).
+    - A channel is considered to be *timed out* if `currentL1Block > channeld_id.timestamp + CHANNEL_TIMEOUT`.
+        - where `currentL1Block` is the L1 block maintained by this stage, which is the latest L1 block whose frames
+          have been added to the channel bank.
+
+> **TODO** There is currently `MAX_CHANNEL_BANK_SIZE`, a notion about the maximum amount of channels we can keep track
+> of.
+> - Is this a semantic detail (i.e. if the batcher opens too many frames, valid channels can be dropped?)
+> - If so, I feel **very strongly** about changing this. This ties us very much to the current implementation.
+> - And it doesn't feel necessary given the channel timeout - if DOS is an issue we can reduce the channel timeout.
```

Comment:
- Note: the size is a BYTE size.
     - And it doesn't matter if we have many channels or few channels with many frames.
- We need this to avoid DOS from malicious batchers.
- Drop full channels at a time.
- Open question: drop old channels or drop new channels.
     - Only relevant in a malicious batcher case, malicious batcher can do damage in both cases.